### PR TITLE
Add BitNet

### PR DESCRIPTION
**Pull request by DustinWang1**
Created BitNet implementation by copying the transformer code and replacing nn.Linear with fused bit linear. Added a "bit" version for the attention module to pair with the BitNet.

### CODE REVIEW

**Comment by yzhangcs** [comment]
@DustinWang1 Hello, thanks for this PR.

What's the diffs between SeptNet and BitNet?

**Comment by DustinWang1** [comment]
Ah I did not mean to add SeptNet into this pull request. Is there any way for you to pull from "init changes"
on your side?

**Comment by yzhangcs** [comment]
@DustinWang1 Feel free to add new commits to delete undesired parts. I will squash them before merging :-)

**Comment by yzhangcs** [comment]
Also It would better to use isort to rearrange imported modules in your new commits

**Comment by yzhangcs** [comment]
@DustinWang1 Another minor reminder, we have updated the attn layer and Xfmr++ impls recently, primarily on Cache update and param inits. Please ensure your code aligns with these latest changes.

**Comment by DustinWang1** [comment]
Thanks for letting me know about isort. There are so many nice utilities out there :0. I've rearranged the imports and synced my changes with the cache update and param inits.

File: fla/models/bitnet/modeling_bitnet.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        self.down_proj = BitLinear(self.intermediate_size, self.hidden_size, bias=False)
+        self.act_fn = ACT2FN[hidden_act]
+
+    def forward(self, x):
+        if self.norm_first:
+            x = rms_norm_linear(x, self.norm.weight, self.norm.bias, self.gate_proj.weight, self.gate_proj.bias)
+        else:
+            x = self.gate_proj(x)
+        gate, y = x.chunk(2, -1)
+        return swiglu_linear(gate, y, self.down_proj.weight, self.down_proj.bias)
```
Comment (yzhangcs) [inline fla/models/bitnet/modeling_bitnet.py]: @DustinWang1 Be careful that once you called `rms_norm_linear` or `swiglu_linear`, the `F.linear` is conducted internally. You did not actually invoke quant layers.

**Comment by yzhangcs** [comment]
@DustinWang1 Thank you for your quick fix. Could you check out my latest comments again.

**Comment by DustinWang1** [comment]
Are talking about the failed check? I'm currently fixing the style errors on my side, but keep in mind that I haven't changed many of the files that flake8 is citing, they were alr in the main repo. There is one error where S is not defined: "fla\ops\generalized_delta_rule\iplr\naive.py:43:9: F821 undefined name 'S'". I'm not sure how this part of the code works, could you take a look?

**Review by yzhangcs — CHANGES_REQUESTED** [review]
@DustinWang1 
https://github.com/DustinWang1/flash-linear-attention/blob/a56e65801f7adbc519edae8c72bfd891a1ddf836/fla/models/bitnet/modeling_bitnet.py#L63-L69

Be careful that once you called `rms_norm_linear` or `swiglu_linear`, the `F.linear` is conducted internally. You did not actually invoke quant layers.

**Comment by DustinWang1** [comment]
I replaced rms_norm_linear with a wrapper of the layer_norm_linear_quant_fn in modules/fused_bitlinear. For the swiglu function, I added an alternate version in modules/activations.py that uses a functional form of BitLinear. The function, bit_linear, is also just a wrapper for layer_norm_linear_quant_fn with rms_norm set to true. Let me know if this fixes the issue.

**Pull request merged**